MerkleMountainRange/src/merklemountainrange.rs

```diff
@@ -0,0 +1,242 @@
+//       /\   \ \        /\   \
+//      /  \   \ \      /  \   \
+//     /\  /\  /\ \    /\  /\  /\
+//    /\/\/\/\/\/\/\  /\/\/\/\/\/\
+//
+// Node 1 has now only been marked as pruned but we cannot remove it as of yet because we still require it to
+// prove node 2. When we prune node 2, the MMR looks as follows
+//
+//                 /\
+//                /  \
+//               /    \
+//              /      \
+//             /        \
+//            /          \
+//           /            \
+//          /\             \
+//         /\ \            /\
+//        /  \ \          /  \
+//       /\   \ \        /\   \
+//      /  \   \ \      /  \   \
+//     /\  /\  /\ \    /\  /\  /\
+//      /\/\/\/\/\/\  /\/\/\/\/\/\
+//
+// Although we have no removed node 1 and node 2 from the MMR we cannot yet remove node 3 as we require node 3 for
+// the proof of node 6. Lets prune 4 and 5.
+//
+//                 /\
+//                /  \
+//               /    \
+//              /      \
+//             /        \
+//            /          \
+//           /            \
+//          /\             \
+//         /\ \            /\
+//        /  \ \          /  \
+//       /\   \ \        /\   \
+//      /  \   \ \      /  \   \
+//         /\  /\ \    /\  /\  /\
+//        /\/\/\/\/\  /\/\/\/\/\/\
+//
+// Now we removed 3 from the MMR
+
+use crate::merklenode::{Hashable, MerkleNode, ObjectHash};
+use std::collections::HashMap;
+
+pub struct MerkleMountainRange<T>
+where T: Hashable
+{
+    mmr: Vec<MerkleNode>,
+    data: HashMap<ObjectHash, T>,
+}
+
+impl<T> MerkleMountainRange<T>
+where T: Hashable
+{
+    /// This function creates a new empty Merkle Mountain Range
+    pub fn new() -> MerkleMountainRange<T> {
+        MerkleMountainRange { mmr: Vec::new(), data: HashMap::new() }
+    }
+
+    /// This function adds an vec of leaf nodes to the mmr.
+    pub fn add_vec(&mut self, objects: Vec<T>) {
+        for object in objects {
+            self.add_single(object);
+        }
+    }
+
+    /// This function adds a new leaf node to the mmr.
+    pub fn add_single(&mut self, object: T) {
+        let node_hash = object.get_hash();
+        let node = MerkleNode::new(node_hash);
+        self.mmr.push(node);
+        if is_node_right((self.mmr.len() - 1) as u64) {
+            self.data.insert(node_hash, object);
+            self.add_single_no_leaf((self.mmr.len() - 1) as u64)
+        }
+    }
+
+    //  //This function adds non leaf nodes, eg nodes that are not directly a hash of data
+    // This is iterative and will continue to add up and till it hits the top, will be a future left child
+    fn add_single_no_leaf(&mut self, index: u64) {
+        let new_hash =
+            self.data.get(&self.mmr[index as usize].hash).unwrap().concat(self.mmr[peer_index(index) as usize].hash);
+
+        let new_node = MerkleNode::new(new_hash);
+        self.mmr.push(new_node);
+        if is_node_right((self.mmr.len() - 1) as u64) {
```

Comment:
The "len() - 1" is used often, would it not make sense creating a member function that returns the last index.